### DISCUSSION TITLE
Continously rotating without jumps between tours

### DISCUSSION THREAD

**Discussion by ethanruiyu** [pmndrs/react-spring#1760] [Support] [answered]
Hi, all.

I found the problem I'm facing in `issues`, but i didn't find the answer.

issues: https://github.com/pmndrs/react-spring/issues/1487

Borrow @korbav codesandbox: https://codesandbox.io/s/twilight-http-l0jm0?file=/src/App.js

As you can see, It will rebound at 0deg - 360deg

https://user-images.githubusercontent.com/81229588/142238768-7326a60b-be53-41ca-928a-d20bb6d265f9.mov

like this, I want to keep control of the angle.

It bothered me for a long time 😭

**Comment by joshuaellis** [comment]
That sandbox uses `react-motion` not `react-spring`.

**Reply by ethanruiyu** [reply]
wtf? forgive me for not looking carefully. I created another one for this: https://codesandbox.io/s/confident-star-fmbz1?file=/src/mobx.js

**Reply by joshuaellis** [reply]
Is there a reason you need it go back to 0 after 360? When im doing something like this, I just keep adding forever. It can handle a rotation above 360.

**Reply by korbav** [reply]
> 
> 
> Is there a reason you need it go back to 0 after 360? When im doing something like this, I just keep adding forever. It can handle a rotation above 360.

Continuously adding will unavoidably conduct to a MAX_VALUE overflow which will set the rotation value to NaN.

**Reply by joshuaellis** [reply]
That's an awfully big number! but thanks for letting me know. I do have an idea anyway to resolve this, but i'm stretched for time. I guess what you could do is check if the new value is going to be `360` and use the `immediate` prop to set it to 0, this would stop any animation. Alternatively this can be done with the `Imperative API` using `start` and `set`.

If you're still struggling, i'll try find time to edit the sandbox to illustrate my point.

**Reply by ethanruiyu** [reply]
The original value is from the hardware sensor, which gives a range of 0-360, and the next value is random (because we can't predict sensor position and orientation).

**Reply by joshuaellis** [reply]
So how do you know if it goes from 358 to 4 they've travelled clockwise? It's a tricky situation because you're trying to go against how CSS works.

If you know the direction you could safely do the following:

- current value is `358`
- next value is `4` (we're going `clockwise`)
- `diff = 6`
- transition to `358 + 6 = 364`
- `onRest` handler animate to 4 with either `immediate = true` or `api.set`
- Spring is at 4 ready for new value between 0-360


You can reverse that calc logic if they're going counter-clockwise. Given the complexities of the animation you're trying to achieve I would use the imperative API.

**Reply by ethanruiyu** [reply]
Sensor reporting frequency is very fast (5Hz), the difference between two consecutive frames is not very big, so 358-4 must be rotating at a small Angle (to determine the direction of rotation).

The idea you mentioned looks great. I could try it, but it might be slow because I'm new to React-Spring😅

**Reply by joshuaellis** [reply]
Write your spring like this:

```js
const [styles, api] = useSpring(() => {
  rotateX: 0,
  // initial-styles
},[])
```

then when you get a state update run a useLayoutEffect/useEffect:

```js
useLayoutEffect(() => {
 // old val could be 358 and new val 4
  const [direction, hasValueDoneRotation] = findDirection(oldVal, newVal)
  // normal change
if(!hasValueDoneRotation){
   api.start({
    rotateX: newRotateX
  })
}
  if(hasValueDoneRotation && direction === 'right'){
    const actualRotationValue = X // this could be 364
    
    api.start({
      rotateX: actualRotationValue,
      onRest: () => {
        api.set({
          // this would be 4
          rotateX: newVal
          })
       }
    })
  }
},[someState])
```

Sorry about the formatting, but it should look something like that.

**Reply by ethanruiyu** [reply]
https://codesandbox.io/s/condescending-euclid-imzmx?file=/src/App.js

This seems to solve the most basic problem, but it doesn't seem to be ideal, and I can't think of a better way.

**Comment by raycaster08** [comment, marked as answer]
Modify the original value based on the difference
https://codesandbox.io/s/flamboyant-ioana-5zems1?file=/src/App.js

**Reply by ethanruiyu** [reply]
nice fix
